bcc/table.go

```diff
@@ -181,56 +212,111 @@ func (table *Table) DeleteAll() error {
+	fd := C.bpf_table_fd_id(table.module.p, table.id)
+
+	return &TableIterator{
+		table: table,
+		fd:    fd,
+	}
+}
+
+// Next looks up the next element and return true if one is available.
+func (it *TableIterator) Next() bool {
+	if it.err != nil {
+		return false
+	}
+
+	if it.key == nil {
+		keySize := C.bpf_table_key_size_id(it.table.module.p, it.table.id)
+
+		key := make([]byte, keySize)
 		keyP := unsafe.Pointer(&key[0])
-		leafP := unsafe.Pointer(&leaf[0])
-		alternateKeys := []byte{0xff, 0x55}
-		res := C.bpf_lookup_elem(fd, keyP, leafP)
-		// make sure the start iterator is an invalid key
-		for i := 0; i <= len(alternateKeys); i++ {
-			if res < 0 {
-				break
-			}
-			for j := range key {
-				key[j] = alternateKeys[i]
+		if res, err := C.bpf_get_first_key(it.fd, keyP, keySize); res != 0 {
+			if !os.IsNotExist(err) {
+				it.err = err
 			}
-			res = C.bpf_lookup_elem(fd, keyP, leafP)
+			return false
 		}
-		if res == 0 {
-			return
-		}
-		keyStr := make([]byte, key_size*8)
-		leafStr := make([]byte, leaf_size*8)
-		keyStrP := (*C.char)(unsafe.Pointer(&keyStr[0]))
-		leafStrP := (*C.char)(unsafe.Pointer(&leafStr[0]))
-		for res = C.bpf_get_next_key(fd, keyP, keyP); res == 0; res = C.bpf_get_next_key(fd, keyP, keyP) {
-			r := C.bpf_lookup_elem(fd, keyP, leafP)
-			if r != 0 {
-				continue
-			}
-			r = C.bpf_table_key_snprintf(mod, table.id, keyStrP, C.size_t(len(keyStr)), keyP)
-			if r != 0 {
-				break
-			}
-			r = C.bpf_table_leaf_snprintf(mod, table.id, leafStrP, C.size_t(len(leafStr)), leafP)
-			if r != 0 {
-				break
-			}
-			ch <- Entry{
-				Key:   string(keyStr[:bytes.IndexByte(keyStr, 0)]),
-				Value: string(leafStr[:bytes.IndexByte(leafStr, 0)]),
+
+		leafSize := C.bpf_table_leaf_size_id(it.table.module.p, it.table.id)
+		leaf := make([]byte, leafSize)
+
+		it.key = key
+		it.leaf = leaf
+	} else {
+		keyP := unsafe.Pointer(&it.key[0])
+		if res, err := C.bpf_get_next_key(it.fd, keyP, keyP); res != 0 {
+			if !os.IsNotExist(err) {
+				it.err = err
 			}
+			return false
+		}
+	}
+
+	keyP := unsafe.Pointer(&it.key[0])
+	leafP := unsafe.Pointer(&it.leaf[0])
+	if res, err := C.bpf_lookup_elem(it.fd, keyP, leafP); res != 0 {
+		it.err = errIterationFailed
+		if !os.IsNotExist(err) {
+			it.err = err
 		}
-	}()
-	return ch
+		return false
+	}
+
+	return true
+}
+
+// Key returns the current key value of the iterator, if the most recent call to Next returned true.
+// The slice is valid only until the next call to Next.
+func (it *TableIterator) Key() []byte {
+	return it.key
+}
+
+// Leaf returns the current leaf value of the iterator, if the most recent call to Next returned true.
+// The slice is valid only until the next call to Next.
+func (it *TableIterator) Leaf() []byte {
+	return it.leaf
+}
+
+// KeyString returns the current key value of the iterator formatted using the bcc-table's key string printer, if the most recent call to Next returned true.
+func (it *TableIterator) KeyString() (string, error) {
```

Comment:
I think users can use `KeyBytesToStr` if they want to, it's not necessary to add this to the iterator itself. Just my 2c.